Extension - Exception observed on unwrapping request with path `//`

## Motivations

- Happened with version `0.9.7`
  - This was introduced in `0.9.7` as a result of a change in the way the Uri is computed for the contained app request due to switching to using connection pools via the Hyper Client
- Panic observed in a deployed env
- The panic at `try_join_all` allowed the different invocations to keep running across Lambda invokes
- A final invoke got stuck for 60 seconds and appears to have handled no requests
- The issue is triggered by a request to the router with path `//`

## Locations

https://github.com/pwrdrvr/lambda-dispatch/blob/eae25adc319d955e0427a09b7f818a654b811fef/extension/src/app_request.rs#L67

https://github.com/pwrdrvr/lambda-dispatch/blob/eae25adc319d955e0427a09b7f818a654b811fef/extension/src/lambda_request.rs#L186

## Acceptance Criteria

- [ ] Determine if the ping does not start until the first request to the Router returns
   - [ ] If this is the case, this needs to be changed to start the pinger earlier, perhaps after a 500 ms delay
   - [ ] The pinger needs to treat all errors as a "goaway" (e.g. cannot connect, non-200 response, etc)
- [ ] Add failing unit test that simulates the error in `app_request.rs`
  - [ ] Propagate back an appropriate error level - continue running if possible, else return back to the router
- [ ] Remove the panic in `lambda_request.rs` and replace it with proper error handling and propagation 

## Logs of Error

```
2024-05-01T14:17:30.827 [ERROR] - PoolId: default, LambdaId: 6d2f4fe0-a19c-4f14-9c15-761799503e41, ChannelId: 2e61af67-13e6-4a31-9fb5-352204493181, ChannelNum: 3 - Error in handle_one_loop: Failed to establish connection to app

thread 'main' panicked at extension/src/app_request.rs:67:66:
called `Result::unwrap()` on an `Err` value: EmptyHost

thread 'main' panicked at extension/src/lambda_request.rs:186:9:
Error in futures::future::try_join_all: task 3250 panicked
```

<img width="1391" alt="image" src="https://github.com/pwrdrvr/lambda-dispatch/assets/5617868/d58dc255-3d96-4458-8736-2737923e7f05">

--- a/extension/src/app_request.rs
+++ b/extension/src/app_request.rs
@@ -374,6 +374,108 @@ mod tests {
   }
 
   #[tokio::test]
+  async fn test_read_until_req_headers_double_slash_path() {
+    let lambda_id = "lambda_id".to_string();
+    let pool_id = "pool_id".to_string();
+    let channel_id = "channel_id".to_string();
+
+    // Start router server
+    let mock_router_server = test_mock_router::setup_router(test_mock_router::RouterParams {
+      request_method: test_mock_router::RequestMethod::GetDoubleSlashPath,
+      channel_non_200_status_after_count: 1,
+      channel_non_200_status_code: StatusCode::CONFLICT,
+      channel_panic_response_from_extension_on_count: -1,
+      channel_panic_request_to_extension_before_start_on_count: -1,
+      channel_panic_request_to_extension_after_start: false,
+      channel_panic_request_to_extension_before_close: false,
+      ping_panic_after_count: -1,
+      listener_type: test_mock_router::ListenerType::Http,
+    });
+
+    // Let the router return right away
+    tokio::spawn(async move {
+      mock_router_server
+        .release_request_tx
+        .send(())
+        .await
+        .unwrap();
+    });
+
+    let channel_url: Uri = format!(
+      "http://localhost:{}/api/chunked/request/{}/{}",
+      mock_router_server.server.addr.port(),
+      lambda_id,
+      channel_id
+    )
+    .parse()
+    .unwrap();
+
+    // Setup our connection to the router
+    let router_client = create_router_client();
+
+    // Create the router request
+    let (_tx, recv) = mpsc::channel::<crate::prelude::Result<Frame<Bytes>>>(32 * 1024);
+    let boxed_body = BodyExt::boxed(StreamBody::new(recv));
+    let req = Request::builder()
+      .uri(&channel_url)
+      .method("POST")
+      .header(hyper::header::DATE, fmt_http_date(SystemTime::now()))
+      .header(hyper::header::HOST, channel_url.host().unwrap())
+      // The content-type that we're sending to the router is opaque
+      // as it contains another HTTP request/response, so may start as text
+      // with request/headers and then be binary after that - it should not be parsed
+      // by anything other than us
+      .header(hyper::header::CONTENT_TYPE, "application/octet-stream")
+      .header("X-Pool-Id", pool_id.to_string())
+      .header("X-Lambda-Id", &lambda_id)
+      .header("X-Channel-Id", &channel_id)
+      .body(boxed_body)
+      .unwrap();
+
+    let res = router_client.request(req).await.unwrap();
+    let (parts, mut res_stream) = res.into_parts();
+
+    let app_endpoint = "http://localhost:3000".parse::<Endpoint>().unwrap();
+
+    // Act
+    let result = read_until_req_headers(
+      app_endpoint,
+      &mut res_stream,
+      &pool_id,
+      &lambda_id,
+      &channel_id,
+    )
+    .await;
+
+    // Assert
+    assert_eq!(
+      parts.headers.get(HeaderName::from_static("content-type")),
+      Some(&hyper::http::HeaderValue::from_static(
+        "application/octet-stream"
+      ))
+    );
+    assert_ok!(&result);
+    let (app_req_builder, goaway, left_over_buf) = result.unwrap();
+    let host_header = app_req_builder.headers_ref().unwrap().get("host");
+    assert_eq!(host_header, None);
+    let test_header = app_req_builder.headers_ref().unwrap().get("test-header");
+    assert_eq!(test_header, Some(&HeaderValue::from_static("foo")));
+    let app_req_uri = app_req_builder.uri_ref().unwrap();
+    assert_eq!(
+      app_req_uri,
+      &Uri::from_static("http://localhost:3000/bananas/no_host_header")
+    );
+    assert_eq!(goaway, false);
+    assert_eq!(left_over_buf.is_empty(), true);
+    assert_eq!(
+      mock_router_server
+        .request_count
+        .load(std::sync::atomic::Ordering::SeqCst),
+      1
+    );
+  }
+
+  #[tokio::test]
   async fn test_read_until_req_headers_go_away_path() {
     let lambda_id = "lambda_id".to_string();
     let pool_id = "pool_id".to_string();

--- a/extension/src/app_request.rs
+++ b/extension/src/app_request.rs
@@ -64,7 +64,7 @@ pub async fn read_until_req_headers(
         // The app_url is only the path
         // Next.js, for one, gives a 308 redirect if you give it `http://localhost:3000/`
         // and it mangles that to `http:/localhost:3000/`
-        let app_url = app_endpoint.url().join(req.path.unwrap()).unwrap();
+        let app_url = format!("{}{}", app_endpoint, req.path.unwrap());
         let app_url = Uri::from_str(app_url.as_str()).unwrap();
 
         let mut app_req_bld = Request::builder()
@@ -457,14 +457,11 @@ mod tests {
     assert_ok!(&result);
     let (app_req_builder, goaway, left_over_buf) = result.unwrap();
     let host_header = app_req_builder.headers_ref().unwrap().get("host");
-    assert_eq!(host_header, None);
+    assert_eq!(host_header, Some(&HeaderValue::from_static("localhost")));
     let test_header = app_req_builder.headers_ref().unwrap().get("test-header");
     assert_eq!(test_header, Some(&HeaderValue::from_static("foo")));
     let app_req_uri = app_req_builder.uri_ref().unwrap();
-    assert_eq!(
-      app_req_uri,
-      &Uri::from_static("http://localhost:3000/bananas/no_host_header")
-    );
+    assert_eq!(app_req_uri, &Uri::from_static("http://localhost:3000//"));
     assert_eq!(goaway, false);
     assert_eq!(left_over_buf.is_empty(), true);
     assert_eq!(

--- a/extension/src/app_request.rs
+++ b/extension/src/app_request.rs
@@ -60,10 +60,10 @@ pub async fn read_until_req_headers(
         );
 
         let left_over_buf = buf[offset..].to_vec();
+
         // The app_url is only the path
         // Next.js, for one, gives a 308 redirect if you give it `http://localhost:3000/`
         // and it mangles that to `http:/localhost:3000/`
-
         let app_url = app_endpoint.url().join(req.path.unwrap()).unwrap();
         let app_url = Uri::from_str(app_url.as_str()).unwrap();
 
@@ -143,6 +143,30 @@ mod tests {
     Request, Uri,
   };
   use tokio_test::assert_ok;
+  use url::Url;
+
+  #[tokio::test]
+  async fn test_url_join() {
+    let base_url = Url::parse("http://example.com").unwrap();
+    let joined_url = base_url.join("http://foo/bar").unwrap();
+    assert_eq!(joined_url.as_str(), "http://foo/bar");
+
+    let base_url = Url::parse("http://example.com:54321").unwrap();
+    let joined_url = base_url.join("").unwrap();
+    assert_eq!(joined_url.as_str(), "http://example.com:54321/");
+
+    let base_url = Url::parse("http://example.com").unwrap();
+    let joined_url = base_url.join("http:///foo").unwrap();
+    assert_eq!(joined_url.as_str(), "http://foo/");
+
+    let base_url = Url::parse("http://example.com").unwrap();
+    // This case results in:
+    //   called `Result::unwrap()` on an `Err` value: EmptyHost
+    // To reproduce, load a router URL with the path `//`
+    let joined_url = base_url.join("//");
+    assert!(joined_url.is_err());
+    assert_eq!(joined_url.err().unwrap(), url::ParseError::EmptyHost);
+  }
 
   #[tokio::test]
   async fn test_read_until_req_headers_valid_req() {

--- a/extension/src/lambda_request.rs
+++ b/extension/src/lambda_request.rs
@@ -167,7 +167,11 @@ impl LambdaRequest {
               );
 
               if err.is_fatal() {
-                return Err(err.into());
+                // Try to clean up the ping task
+                self.cancel_token.cancel();
+                let _ = tokio::time::timeout(std::time::Duration::from_secs(1), ping_task).await;
+
+                return Err(err);
               }
 
               // Error is not fatal so just use it as an exit reason
@@ -177,16 +181,25 @@ impl LambdaRequest {
         }
       }
       Err(err) => {
+        // try_join_all returns JoinError when one of the futures panics
+        // If a future has panicked we are probably in a bad state and should exit
         log::error!(
           "LambdaId: {} - run - Error in futures::future::try_join_all: {:?}",
           self.lambda_id,
           err
         );
-        // TODO: Capture the channel exit error and return the worst one we find
-        panic!(
-          "LambdaId: {} - run - Error in futures::future::try_join_all: {}",
-          self.lambda_id, err
-        );
+
+        // Set the goaway signal so other tasks stop
+        self.goaway_received.store(true, Ordering::Release);
+
+        // Try to clean up the ping task
+        self.cancel_token.cancel();
+        let _ = tokio::time::timeout(std::time::Duration::from_secs(1), ping_task).await;
+
+        // This is a bit of a lie: we don't know if the app connection is unreachable
+        // but we do know we're in an non-deteministic state and that this
+        // Lambda should exit and be recreated
+        return Err(LambdaRequestError::AppConnectionUnreachable);
       }
     }
 

--- a/extension/src/router_channel.rs
+++ b/extension/src/router_channel.rs
@@ -461,10 +461,18 @@ impl RouterChannel {
         }
       }
       Err(err) => {
-        panic!(
-          "PoolId: {}, LambdaId: {}, ChannelId: {} - Error in futures::future::try_join_all: {}",
-          self.pool_id, self.lambda_id, self.channel_id, err
+        // try_join_all will return an error if either of the futures panics
+
+        log::error!(
+          "PoolId: {}, LambdaId: {}, ChannelId: {} - pickup-request - Error in futures::future::try_join_all: {}",
+          self.pool_id,
+          self.lambda_id,
+          self.channel_id,
+          err
         );
+
+        // We don't have details about which future failed, so we return a generic error
+        return Err(LambdaRequestError::ChannelErrorOther);
       }
     }
 

--- a/extension/src/test_mock_router.rs
+++ b/extension/src/test_mock_router.rs
@@ -18,6 +18,7 @@ use tokio::{io::AsyncWriteExt, sync::mpsc::Sender};
 #[derive(Clone, Copy, PartialEq)]
 pub enum RequestMethod {
   Get,
+  GetDoubleSlashPath,
   GetNoHost,
   GetQuerySimple,
   GetQueryEncoded,
@@ -193,6 +194,9 @@ pub fn setup_router(params: RouterParams) -> RouterResult {
 
                                 let data = "a".repeat(10 * 1024);
                                 tx.write_all(data.as_bytes()).await.unwrap();
+                            } else if params.request_method == RequestMethod::GetDoubleSlashPath {
+                                let data = b"GET // HTTP/1.1\r\nHost: localhost\r\nTest-Header: foo\r\n\r\n";
+                                tx.write_all(data).await.unwrap(); 
                             } else if params.request_method == RequestMethod::GetQuerySimple {
                                 let data = b"GET /bananas_query_simple?cat=dog&frog=log HTTP/1.1\r\nHost: localhost\r\nTest-Header: foo\r\n\r\n";
                                 tx.write_all(data).await.unwrap();
@@ -214,6 +218,7 @@ pub fn setup_router(params: RouterParams) -> RouterResult {
                             } else if params.request_method == RequestMethod::GetEnormousHeaders {
                                 let data = b"GET /bananas/enormous_headers HTTP/1.1\r\nHost: localhost\r\n";
                                 tx.write_all(data).await.unwrap();
+            
 
                                 let header_value = "a".repeat(1024 * 31); // Each header value is 31 KB
                                 for i in 0..4 {

